DHCPv4 Relay Agent High Level Design

ashutosh-agrawal opened the pull request:
This document provides a high-level design for SONiC's new DHCPv4 relay agent, which addresses limitations of the existing ISC-DHCP implementation. The document outlines the architecture, new features and configuration of the new design, and also addresses the backward compatibility concerns related to transition from isc-dhcp.

Code PRs:
[sonic-dhcp-relay](https://github.com/sonic-net/sonic-dhcp-relay/pull/67)
[yang-model](https://github.com/sonic-net/sonic-buildimage/pull/22338)
[sonic-buildimage](https://github.com/sonic-net/sonic-buildimage/pull/22486)
[sonic-swss-common](https://github.com/sonic-net/sonic-swss-common/pull/1017)
[Testplan HLD](https://github.com/sonic-net/sonic-mgmt/pull/19215)
[sonic-mgmt test](https://github.com/sonic-net/sonic-mgmt/pull/19198)
[CLIs](https://github.com/sonic-net/sonic-buildimage/pull/23242)

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

yaqiangz commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+The existing DHCP monitor will be enhanced in-order to support monitoring of DHCPv4 packets being handled by the new implementation.
+
+## Dual-Tor Support
+In a dual-TOR (Top-of-Rack) architecture, it's possible for DHCP request packets from a client to reach the server through one TOR, while the server's response might arrive at the peer TOR. If the peer TOR has its link in standby mode, it won't be able to relay the response back to the originating client. This issue can be addressed by configuring the DHCPv4 relay with the link-selection and sour ...
```
Could you please add detail of dualtor? i.e. What is the value of giaddr of tx packets. How dual tor proceed rx packets? Seems you plan to use vlan-id in circuit id, if customizing option 82 is feasible, can we fallback to giaddr or yiddr if hostname is not set?
@lolyu I think dst ip of reply packets sent by DHCP server would be giaddr in DHCP packets. Is it acceptable to set giaddr as vlan ip in dualtor? Then DHCP server would send DHCP reply packets with dst ip set as vlan ip.

ashutosh-agrawal replied:
Giaddr of the the packets forwarded by the relay will be dependent on the DHCP relay configuration. If the link-selection and source-interface options are set as described below, giaddr will be set to the Loopback IP of the relay.
For the received packet, if option 82 doesn't have the necessary information to identify the downstream VLAN interface , then giaddr will be used to identify it. We can add yiddr also as a secondary fallback option if it's needed for dual-tor operation. Please let us know.

yaqiangz commented:
Hi @lolyu could you please help to check the dual tor part?

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+
+- **R9:** Support configuration of the source interface to specify the source IP address for relayed packets.
+
+- **R10:** Support handling of DHCP packets that have already been relayed by other agents. Three options are available:
```
It's identified by the presence of Option 82. If option 82 is present and the giaddr is zero, packet should be dropped.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+
+A DHCP relay agent is an essential component in networks where clients and DHCP servers are on different subnets, enabling communication by forwarding DHCP messages across subnet boundaries. Here's how a DHCP relay agent manages the four basic DHCP messages:
+
+- **DHCP Discover:** When a DHCP client broadcasts a Discover message to locate available DHCP servers, the message is confined to the local subnet. The DHCP relay agent, typically configured on a router or Layer 3 switch, intercepts this broadcast. It encapsulates the Discover message in a unicast packet directed to the DHCP server and updates the 'giaddr' (gateway IP address) field with its own ...
```
Yes, we will support DHCP NAK/Release/Inform/Decline messages as well. They will be forwarded like other DHCP messages based on the Opcode in the DCHP header. We do have a separate counter for each of the 8 DHCP message types.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+- **Parse Relay Sub-options:** The relay agent parses and extracts all relay sub-options from the packet, which may include information such as Circuit-ID, Remote-ID, and other relevant details.
+- **Identify Downstream Interface**
+		- Using Circuit-ID: If the packet contains a Circuit-ID sub-option, the relay agent uses the VLAN portion of the Circuit-ID to identify the appropriate downstream interface through which to forward the packet to the client. This method provides a direct and efficient way to determine the correct path for the packet.
+		- Fallback to giaddr: If the Circuit-ID is absent, the relay agent falls back on the giaddr (gateway IP address) field. It loops through each interface IP and compares it with the giaddr value to identify the corresponding downstream interface. This ensures that even without a Circuit-ID, the packet can still be accurately directed to the client.
```
Yes, the design will fallback to giaddr if either circuit-id is absent OR the circuit-id doesn't have the necessary information to identify the client-side interface for forwarding the packet.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+- **BOOTREQUEST:** If the Opcode indicates a BOOTREQUEST, the packet is processed through the Client-to-Server packet handling pathway. This involves forwarding the client's DHCP request to the appropriate DHCP server, potentially adding relay information options as configured.
+- **BOOTREPLY :** If the Opcode is not a BOOTREQUEST, the packet is assumed to be a response from a DHCP server and is processed through the Server-to-Client packet handling pathway. This involves relaying the DHCP server's response back to the client on the originating subnet.
+
+#### Client->Server Packet Handling
```
Yes, we will ensure that IP and UDP checksum are validated for the received packet.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+
+
+        | Subopt | Len | Circuit ID |
+        | -------|-----|-------------------------------|
```
We have looked into the patches but I'm a bit unclear about the use cases that require a customization of circuit/remote-id. If there are specific use cases that are not covered by carrying {hostname, physical-interface-alias, vlan} in the circuit-id please let us know. Based on this [jinja template](https://github.com/sonic-net/sonic-buildimage/blob/master/dockers/docker-dhcp-relay/dhcpv4-relay.agents.j2) SONiC is currently carrying {hostname, int-alias} in the circuit-id and we are proposing a superset of it by also including vlan-id in it.

yaqiangz replied:
Why vlan information is needed here? It looks like duplicated with interface name. From interface name, we could get vlan id. Also adding vlan into option 82 is a breaking change, the reason why I want customizing it is to make sure it could be same as previous. We don't know how all the upstream (DHCP SEVER) will use this field or whether they will full match the Option 82.

ashutosh-agrawal replied:
If client is connected to an access port, we can get vlan-id from the interface name as you mentioned. But if the client is attached to a trunk port, there could be multiple vlans configured on the same port.

In my research, I found several platforms that include vlan-id in circuit-id but they also provide various config knobs to manage the functionality. Here is an example:

https://www.juniper.net/documentation/us/en/software/junos/cli-reference/topics/ref/statement/circuit-id-edit-forwarding-options.html

I'm considering two possible solutions to your concerns regarding the backward compatibility:

1) Include vlan-id in the circuit-id only if the packet arrived on a trunk port.
2) Introduce a new feature flag in config_db.json to let the users decide between the two modes. This will be a global flag and the default will be to not include the vlan-id.

I'm open to any other suggestions from the community as well.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+                leaf source_interface {
+                    description "Used to determine the source IP address of the relayed packet";
+                    type union {
+                        type leafref {
```
Do you see any downside to using interface here given that we plan to listen to Redis updates for any changes to interface IP?

Regarding multiple IP addresses, our plan is to pick an IP that doesn't have the "secondary" attribute set.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
````diff
@@ -0,0 +1,691 @@
+
+- **Usage**
+```CMD
+sudo config vlan dhcp_relay ipv4 (add | del) <vlan_id> <dhcp_relay_destination_ips>
````
Sure, we can remove it based on the community feedback.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
````diff
@@ -0,0 +1,691 @@
+    -  Add dhcpv4 relay configuration on a VLAN<br>
+
+        ```CMD
+        root@sonic:/home/cisco# config dhcpv4-relay add -h
````
We are currently generating CLIs directly from the yang-model, so the cli syntax reflects the container names in the yang model.

ashutosh-agrawal replied:
We can add manual CLIs you have requested as well.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
````diff
@@ -0,0 +1,691 @@
++-------------+----------------------+
+```
+
+**New CLI to show dhcpv4 relay configuration**
````
Ack

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
````diff
@@ -0,0 +1,691 @@
+
+    * Show counters syntax
+        ```
+        # show dhcp4relay_counters --help
````
Okay, we will add `show dhcp_relay ipv4 counters` to align with dhcpv6.

yaqiangz commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+## Dual-Tor Support
+In a dual-TOR (Top-of-Rack) architecture, it's possible for DHCP request packets from a client to reach the server through one TOR, while the server's response might arrive at the peer TOR. If the peer TOR has its link in standby mode, it won't be able to relay the response back to the originating client. This issue can be addressed by configuring the DHCPv4 relay with the link-selection and sour ...
+
+By enabling the link-selection option, the DHCP relay will use the interface specified by the source-interface option to populate the giaddr field in the packet. When the loopback interface is set as the source-interface, the DHCP request packet sent from the client will have the loopback IP of the originating TOR in the giaddr field. If the DHCP response arrives at the peer TOR, which is in stan ...
```
Would link-selection be enabled by default? If it is disabled, what's the value of giaddr would be set in dualtor?

ashutosh-agrawal replied:
link-selection needs to be enabled for the dual-tor if the intent to forward reply packets through the originating tor only.
If link-selection is disabled, giaddr will carry the VLAN interface IP.

mssonicbld commented:
/azp run

selva-nexthop commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+
+- **R6:** Support DHCP relay in networks using Static Anycast Gateways(SAG), where multiple switches route packets using a common gateway address.
+
+- **R7:** Support relaying DHCP packets between the relay agent and DHCP server on point-to-point links configured as unnumbered interfaces. [RFC 5549](https://datatracker.ietf.org/doc/html/rfc5549)
```
the testplan needs to cover unnumbered interfaces. Validation must include topology changes where relay interfaces transition between numbered/unnumbered states

selva-nexthop commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+        | Vlan10      |    3  |    3  |    1  |    1  |
+        +-------------+-------+-------+-------+-------+
+
+# Yang Model
```
it wil be great if we add a migration guide section on how existing configuraitons map to the new YANG model.

ashutosh-agrawal replied:
Added a table show how the existing configuration will migrate to the new yang model.

selva-nexthop replied:
Thank you.

lolyu reviewed (dismissed):
The dualtor part LGTM

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+
+When a DHCP client broadcasts a request for an IP address, the nature of broadcast traffic confines this request to its local subnet. If the DHCP server is situated on a different subnet, the request would not reach the server without intervention. This is where a network switch with DHCP relay capabilities is essential. The switch intercepts the broadcast DHCP request and forwards it as a unicas ...
+
+The relay process involves adding crucial information, such as the 'giaddr' (gateway IP address) field in the packet, which aids the DHCP server in identifying the subnet from which the request originated. Once the DHCP server assigns an IP address, the switch relays the server's response back to the client, completing the DHCP transaction. This relay functionality is crucial for ensuring efficie ...
```
This new design doesn't change how packet forwarding will work during failover. In Dual TOR case, DHCP ack packets are expected to carry loopback IP of the originating TOR in the destination IP field and will be forwarded based on the route table programming on transit switches like standby TOR. There is no change in the forwarding behavior between the current design and the new design.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+
+- **R8:** Support the DHCP Relay Information option (Option 82), which allows network administrators to include additional information in relayed DHCP messages. Following relay sub-options will be supported.
+
+    - **Circuit ID and Remote ID sub-option**: Allows relay agent to insert specific information related to location of the client in the relay requests. **These sub-options will be inserted by default into every relay packet.
```
I understand the general benefit of providing more flexibility but didn't understand the issue with existing option 82 data. Can you please elaborate on it with some examples.

Also, in the current design, option 82 is enabled by default and there is no field in either config_db or CLI to disable it.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+
+    - **Circuit ID and Remote ID sub-option**: Allows relay agent to insert specific information related to location of the client in the relay requests. **These sub-options will be inserted by default into every relay packet.
+
+    * **Link Selection sub-option** [RFC 3527](https://datatracker.ietf.org/doc/html/rfc3527): Specifies the subnet on which the DHCP client resides, allowing servers to allocate addresses from the correct pool.
```
Sorry, I didn't understand the question. There is a section later in the document that describes the format of link-selection sub-option. Any server functionality is out-of-scope for this document and we are not expecting any changes there.

selva-nexthop replied:
OK.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+
+# Detailed Design
+
+DHCPv4 relay process will run in the DHCP container along with DHCPv6 processes and DhcpMon. A single instance of the process will handle DHCPv4 relay functionality of all the VLANs that are configured. This process will listen to Redis for all the necessary configuration updates and will not require restarting of the container. The design is split into 3 sub-modules and the following diagram pro ...
```
There is a performance section towards the end of the document that briefly touches on this issue. There is no specific logic in the worker thread to prevent starvation and it's expected to be achieved through a combination of CoPP in NPU and DoS mitigation feature described [here](https://github.com/sonic-net/SONiC/blob/master/doc/Dhcp_Mitigation/DHCP%20Mitigation.md).

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+
+    - Discard: Discard the incoming packet (default).
+
+    * Append: Append its own set of relay options to the packet, leaving the incoming options intact. If the length of relay agent information exceeds the max limit of 255 bytes, the packet is discarded.
```
Yes, there is a per-VLAN counter for dropped packets. Please search for "Drp" in the HLD for details.

AnantKishorSharma commented:
Hi @yaqiangz , @saiarcot895 , @lguohan , could you please create the new repo mentioned in this HLD( https://github.com/sonic-net/sonic-dhcpv4-relay) to open code PR?

jasper-hou-micas commented on doc/DHCPv4_relay/images/DHCPv4_Relay_Basic_Flow.png:
Maybe detailing the type of OFFER packet(unicast) is more friendly

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+
+When a DHCP client broadcasts a request for an IP address, the nature of broadcast traffic confines this request to its local subnet. If the DHCP server is situated on a different subnet, the request would not reach the server without intervention. This is where a network switch with DHCP relay capabilities is essential. The switch intercepts the broadcast DHCP request and forwards it as a unicas ...
+
+The relay process involves adding crucial information, such as the 'giaddr' (gateway IP address) field in the packet, which aids the DHCP server in identifying the subnet from which the request originated. Once the DHCP server assigns an IP address, the switch relays the server's response back to the client, completing the DHCP transaction. This relay functionality is crucial for ensuring efficie ...
```
Yes, MC-LAG is supported and listed in the requirements section.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+
+When a DHCP client broadcasts a request for an IP address, the nature of broadcast traffic confines this request to its local subnet. If the DHCP server is situated on a different subnet, the request would not reach the server without intervention. This is where a network switch with DHCP relay capabilities is essential. The switch intercepts the broadcast DHCP request and forwards it as a unicas ...
+
+The relay process involves adding crucial information, such as the 'giaddr' (gateway IP address) field in the packet, which aids the DHCP server in identifying the subnet from which the request originated. Once the DHCP server assigns an IP address, the switch relays the server's response back to the client, completing the DHCP transaction. This relay functionality is crucial for ensuring efficie ...
```
If mutiple IP addresses are configured on a SVI, it's assumend that all but one of the IPs are marked as secondary. The dhcp-relay will pick the primary (not marked as secondary) IP address as giaddr.

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,691 @@
+## Dual-Tor Support
+In a dual-TOR (Top-of-Rack) architecture, it's possible for DHCP request packets from a client to reach the server through one TOR, while the server's response might arrive at the peer TOR. If the peer TOR has its link in standby mode, it won't be able to relay the response back to the originating client. This issue can be addressed by configuring the DHCPv4 relay with the link-selection and sour ...
+
+By enabling the link-selection option, the DHCP relay will use the interface specified by the source-interface option to populate the giaddr field in the packet. When the loopback interface is set as the source-interface, the DHCP request packet sent from the client will have the loopback IP of the originating TOR in the giaddr field. If the DHCP response arrives at the peer TOR, which is in stan ...
```
if link-selection is disabled, giaddr in bootp request will be set to the IP address of the primary client side interface.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,696 @@
+# Design Considerations
+
+## Co-existence with ISC-DHCP Code
+To facilitate a smooth transition for users migrating from ISC-DHCP to the new SONiC-DHCPv4-Relay, both implementations will coexist within the SONiC codebase for a period of time. A new configuration flag, has_sonic_dhcpv4_relay, will be introduced in the dhcp_relay feature within the config-db. This flag will allow users to select either the ISC-DHCP or the SONiC-DHCPv4-Relay implementation, wi ...
```
Yes, there will be an automatic translation of config from VLAN table to DHCPV4_RELAY table in the config_db.

ashutosh-agrawal commented:
@prsunny @yaqiangz @selva-nexthop @jasper-hou-micas @zhangyanzhao I've addressed all the review comments on the HLD. Also, code PRs are open and I've added the link to code PRs above. Appreciate your help in quick approval of the HLD as this feature is included in the 202505 SONiC Planing table.

yaqiangz reviewed (approved):
LGTM

ashutosh-agrawal commented:
@prsunny @zhangyanzhao This HLD has not approvals from @yaqiangz from MSFT and @selva-nexthop for nexthop.ai. Also @lolyu has approved the dual-TOR specific section of the design. Approciate your help in merging the HLD and assigning approvars for the code PRs.

Code PRs:
https://github.com/sonic-net/sonic-dhcp-relay/pull/67
https://github.com/sonic-net/sonic-buildimage/pull/22338
https://github.com/sonic-net/sonic-buildimage/pull/22486
https://github.com/sonic-net/sonic-swss-common/pull/1017

mssonicbld commented:
/azp run

ashutosh-agrawal commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,736 @@
+
+### 6. Detailed Design
+
+DHCPv4 relay process will run in the DHCP container along with DHCPv6 processes and DhcpMon. A single instance of the process will handle DHCPv4 relay functionality of all the VLANs that are configured. This process will listen to Redis for all the necessary configuration updates and will not require restarting of the container. The design is split into 3 sub-modules and the following diagram pro ...
```
Fixed. Thanks.

AnantKishorSharma commented on doc/DHCPv4_relay/DHCPv4-relay-agent-High-Level-Design.md:
```diff
@@ -0,0 +1,739 @@
+
+DHCPv4 relay process will run in the dhcp_relay container along with DHCPv6 processes and DhcpMon. A single instance of the process will handle DHCPv4 relay functionality of all the VLANs that are configured. This process will listen to Redis for all the necessary configuration updates and will not require restarting of the container. The design is split into 3 sub-modules and the following diagr ...
+
+<div align="center"> <img src=images/DHCPv4_Relay_sequence_diagram.png width=700 /> </div>
```
This diagram needs to be updated with the latest changes?
